File: archive/tokenapp/static/fitbit_login.js
/*
 * Script that handles the Fitbit login process
 * Created by: Jasmine Jones
 * Last modified by: Abigail Franz on 1/19/2018
 */

var fbTok;

// If this boolean is set to false, the callback will just be localhost
var usingCsServer = true;
var isJasmine = false;

var clientID = '228N2H';
console.log('clientID = ' + clientID);

homeURL = '';
if (usingCsServer) {
	homeURL = 'https://powertoken.grouplens.org';
} else {
	homeURL = 'http://localhost:5000';
}
callback = homeURL + '/fb_login';
landingPage = homeURL + '/home';

var authURI = 'https://www.fitbit.com/oauth2/authorize';
var authTokenReq = 'https://api.fitbit.com/oauth2/token';

// CHECK for returned code i.e. url#code, if no hash there, then redirect to auth site
if (!window.location.hash) {
	window.location.replace(authURI+'?response_type=token&client_id='+clientID+'&redirect_uri='+callback
		+'&scope=activity%20nutrition%20heartrate%20location%20nutrition%20profile%20settings%20sleep%20social%20weight');
} else {
	var fragmentQueryParameters = {};
	window.location.hash.slice(1).replace(
		new RegExp("([^?=&]+)(=([^&]*))?", "g"),
		function($0, $1, $2, $3) { fragmentQueryParameters[$1] = $3; }
	);
    
	fbTok = fragmentQueryParameters.access_token;
}

var processResponse = function(response) {
	if (!response.ok) {
		console.log("response = " + response);
		throw new Error('Request failed' + response);
	} else {
		window.location.href = landingPage;
	}
}

// SEND token to server
if (fbTok) {
    console.log("redirecting back to app")
    fetch( homeURL+'/result', {
        method: 'POST', 
	   mode: 'no-cors',
	   body: JSON.stringify({tok: fbTok})
    })
    .then(processResponse)
}

File: fitbit_login.py
from flask import Flask, request, redirect
from flask.sessions import SecureCookieSession

app = Flask(__name__)
session = SecureCookieSession()
session.permanent = True
session.modified = True
print("Created app and session objects.")

client_id = "22CV5V"
callback_url = "https://powertoken.grouplens.org/fitbit_result"
auth_url = "https://www.fitbit.com/oath2/authorize"
redirect_url = "{}?response_type=code&client_id={}&redirect_uri={}&scope=activity".format(
	auth_url, client_id, callback_url)

@app.route("/")
@app.route("/index")
def index():
	return redirect(url_for("fitbit_login"))

@app.route("/fitbit_login", methods=["GET", "POST"])
def fitbit_login():
	if request.method == "GET":
		session["username"] = "Jill Pole"
		session.modified = True
		return redirect(redirect_url)

@app.route("/fitbit_result")
def fitbit_result():
	fb_token = request.args.get("access_token")
	print(fb_token)
	return redirect(url_for("index"))

if __name__ == "__main__":
	app.run(debug=True)